packages/backend/src/modules/tracked-txs/modules/liveness/indexers/AnomaliesIndexer.ts

```diff
@@ -0,0 +1,265 @@
+
+    await this.$.anomaliesRepository.addOrUpdateMany(anomalies)
+
+    return unixTo.toNumber()
+  }
+
+  override async invalidate(targetHeight: number): Promise<number> {
+    // no need to remove data
+    // safeHeight will be updated to this value
+    return await Promise.resolve(targetHeight)
+  }
+
+  async getAnomalies(to: UnixTime) {
+    const anomalies: AnomaliesRecord[] = []
+
+    const configurations =
+      await this.$.indexerService.getSavedConfigurations<TrackedTxConfigEntry>(
+        'tracked_txs_indexer',
+      )
+
+    const projectsToSync = getProjectsToSync(this.$.projects, configurations)
+
+    // we need data from 2 * SYNC_RANGE past days to calcualate standard deviation
+    const deviationRange = to.add(-1 * this.SYNC_RANGE * 2, 'days')
+
+    for (const project of projectsToSync) {
+      const livenessRecords =
+        await this.$.livenessRepository.getWithSubtypeByProjectIdsWithinTimeRange(
+          project.projectId,
+          deviationRange,
+          to,
+        )
+
+      if (livenessRecords.length === 0) {
+        this.logger.debug('No records found for project', {
+          projectId: project.projectId,
+        })
+        continue
+      }
+
+      this.logger.debug('Liveness records loaded', {
+        projectId: project.projectId,
+        count: livenessRecords.length,
+      })
+
+      const [batchSubmissions, stateUpdates, proofSubmissions] =
+        groupByType(livenessRecords)
+
+      const test = this.detectAnomalies(
+        project.projectId,
+        'batchSubmissions',
+        batchSubmissions,
+        to,
+      )
+
+      anomalies.push(...test)
+
+      anomalies.push(
+        ...this.detectAnomalies(
+          project.projectId,
+          'stateUpdates',
+          stateUpdates,
+          to,
+        ),
+      )
+
+      anomalies.push(
+        ...this.detectAnomalies(
+          project.projectId,
+          'proofSubmissions',
+          proofSubmissions,
+          to,
+        ),
+      )
+    }
+
+    return anomalies
+  }
+
+  detectAnomalies(
+    projectId: ProjectId,
+    subtype: TrackedTxsConfigSubtype,
+    livenessRecords: LivenessRecordWithSubtype[],
+    to: UnixTime,
+  ): AnomaliesRecord[] {
+    if (livenessRecords.length === 0) {
+      return []
+    }
+
+    // if the oldest record is newer than 2 * SYNC_RANGE -1 we can't calculate anomalies
+    const lastRecord = livenessRecords.at(-1)
+    if (
+      lastRecord?.timestamp.gt(to.add(-1 * (2 * this.SYNC_RANGE - 1), 'days'))
+    )
+      return []
+
+    const anomalies: AnomaliesRecord[] = []
+
+    // add virtual point to calculate ongoing anomalies
+    livenessRecords.unshift({
+      timestamp: to,
+      subtype,
+    })
+
+    const intervals = calculateIntervals(livenessRecords)
+
+    const lastIndex = intervals.findIndex((interval) =>
+      interval.record.timestamp.lte(to.add(-1 * this.SYNC_RANGE, 'days')),
+    )
+
+    const { means, stdDevs } = this.calculate30DayRollingStats(
```

Comment:
whats stdDevs?

Reply:
Renamed it to `stdDeviations` to be more readable